rex/src/map.rs

```diff
@@ -211,36 +156,178 @@ where
+        let map_kptr = unsafe { core::ptr::read_volatile(&self.kptr) };
+        if unlikely(map_kptr.is_null()) {
+            return Err(EINVAL as i32);
+        }
+
+        termination_check!(unsafe {
+            to_result!(ffi::bpf_ringbuf_output(
+                map_kptr,
+                data.as_ptr() as *const (),
+                data.len() as u64,
+                flags
+            ))
+        })
+    }
+
+    /// Queries the amount of data not yet consumed.
+    ///
+    /// Returns `None` is `self` is not a valid ring buffer.
+    pub fn available_bytes(&'static self) -> Option<u64> {
+        let map_kptr = unsafe { core::ptr::read_volatile(&self.kptr) };
+        if unlikely(map_kptr.is_null()) {
+            return None;
+        }
+
+        termination_check!(unsafe {
+            Some(ffi::bpf_ringbuf_query(map_kptr, BPF_RB_AVAIL_DATA as u64))
+        })
+    }
+
+    /// Queries the size of ring buffer.
+    ///
+    /// Returns `None` is `self` is not a valid ring buffer.
+    pub fn size(&'static self) -> Option<u64> {
+        let map_kptr = unsafe { core::ptr::read_volatile(&self.kptr) };
+        if unlikely(map_kptr.is_null()) {
+            return None;
+        }
+
+        termination_check!(unsafe {
+            Some(ffi::bpf_ringbuf_query(map_kptr, BPF_RB_RING_SIZE as u64))
+        })
+    }
+
+    /// Queries the consumer position, which may wrap around.
+    ///
+    /// Returns `None` is `self` is not a valid ring buffer.
+    pub fn consumer_position(&'static self) -> Option<u64> {
+        let map_kptr = unsafe { core::ptr::read_volatile(&self.kptr) };
+        if unlikely(map_kptr.is_null()) {
+            return None;
+        }
+
+        termination_check!(unsafe {
+            Some(ffi::bpf_ringbuf_query(map_kptr, BPF_RB_CONS_POS as u64))
+        })
+    }
+
+    /// Queries the Producer(s) position which may wrap around.
+    ///
+    /// Returns `None` is `self` is not a valid ring buffer.
+    pub fn producer_position(&'static self) -> Option<u64> {
+        let map_kptr = unsafe { core::ptr::read_volatile(&self.kptr) };
+        if unlikely(map_kptr.is_null()) {
+            return None;
+        }
+
+        termination_check!(unsafe {
+            Some(ffi::bpf_ringbuf_query(map_kptr, BPF_RB_PROD_POS as u64))
+        })
+    }
+}
+
+pub struct RexRingBufEntry<'a> {
+    data: &'a mut [u8],
+}
+
+impl RexRingBufEntry<'_> {
+    /// Consumes the reserved payload and submits it to the ring buffer.
+    ///
+    /// If [`crate::linux::bpf::BPF_RB_NO_WAKEUP`] is specified in `flags`,
+    /// no notification of new data availability is sent.
+    /// If [`crate::linux::bpf::BPF_RB_FORCE_WAKEUP`] is specified in `flags`,
+    /// notification of new data availability is sent unconditionally.
+    /// If `0` is specified in `flags`, an adaptive notification of new data
+    /// availability is sent.
+    ///
+    /// This method always succeeds.
+    pub fn submit(self, flags: u64) {
```

Comment:
Since we only have 3 flags, should we create separate `submit` and `discard` variants for them?